quinn/src/lib.rs

```diff
@@ -481,137 +414,91 @@ impl EndpointInner {
 
     fn drive_send(&mut self) -> Result<bool, io::Error> {
         let mut blocked = false;
-        if let Some(ref x) = self.outgoing {
-            match self.socket.poll_send(&x.destination, x.ecn, &x.packet) {
-                Ok(Async::Ready(_)) => {
-                    self.outgoing = None;
-                }
+        while let Some(t) = self.outgoing.pop_front() {
+            match self.socket.poll_send(&t.destination, t.ecn, &t.packet) {
+                Ok(Async::Ready(_)) => {}
                 Ok(Async::NotReady) => {
+                    self.outgoing.push_front(t);
                     blocked = true;
                 }
                 Err(ref e) if e.kind() == io::ErrorKind::PermissionDenied => {
+                    self.outgoing.push_front(t);
                     blocked = true;
                 }
                 Err(e) => {
                     return Err(e);
                 }
             }
-        }
-        Ok(blocked)
-    }
-
-    fn drive_transmit(&mut self, now: Instant) -> Result<bool, io::Error> {
-        let mut sent = 0;
-        while let Some(x) = self.inner.poll_transmit(now) {
-            match self.socket.poll_send(&x.destination, x.ecn, &x.packet) {
-                Ok(Async::Ready(_)) => {}
-                Ok(Async::NotReady) => {
-                    self.outgoing = Some(x);
-                    break;
-                }
-                Err(ref e) if e.kind() == io::ErrorKind::PermissionDenied => {
-                    self.outgoing = Some(x);
-                    break;
-                }
-                Err(e) => {
-                    return Err(e);
-                }
-            }
-            sent += 1;
-            if sent >= IO_LOOP_BOUND {
+            if blocked {
                 return Ok(true);
```

Comment:
We should return true iff `IO_LOOP_BOUND` iterations were completed, and then or that into `keep_going` in the main loop, rather than discarding the value as is currently being done.

Reply:
Done.